lib/elastic_apm/timed_worker.rb

```diff
@@ -0,0 +1,113 @@
+# frozen_string_literal: true
+
+module ElasticAPM
+  # @api private
+  class TimedWorker
+    include Log
+
+    SLEEP_INTERVAL = 0.1
+
+    # @api private
+    class StopMsg
+    end
+
+    # @api private
+    class ErrorMsg
+      def initialize(error)
+        @error = error
+      end
+
+      attr_reader :error
+    end
+
+    def initialize(config, messages, pending_transactions, adapter)
+      @config = config
+      @messages = messages
+      @pending_transactions = pending_transactions
+      @adapter = adapter
+
+      @last_sent_transactions = Time.now.utc
+
+      @serializers = Struct.new(:transactions, :errors).new(
+        Serializers::Transactions.new(config),
+        Serializers::Errors.new(config)
+      )
+    end
+
+    attr_reader :config, :messages, :pending_transactions
+
+    def run_forever
+      loop do
+        run_once
+        sleep SLEEP_INTERVAL
+      end
+    end
+
+    def run_once
+      collect_and_send_transactions if should_flush_transactions?
+      process_messages
+    end
+
+    private
+
+    def process_messages
+      while (msg = messages.pop(true))
+        case msg
+        when ErrorMsg
+          post_error msg
+        when StopMsg
+          # empty collected transactions before exiting
+          collect_and_send_transactions
+
+          Thread.exit
```

Comment:
Rather than exiting here, would it perhaps be better to mark a flag and exit when the empty queue exception raises? Otherwise you might still have messages in the queue.

Reply:
Good idea!